### DISCUSSION TITLE
ipsec pool command report "opening database failed." when using attr-sql plugin sqlite

### DISCUSSION THREAD

**Discussion by shenminjie666** [strongswan/strongswan#2538] [Q&A] [answered]
Hello all

I am currently using strongswan 5.8.4, centos 7.8 x86_64
I build from code like this:

configure --enable-eap-identity --enable-eap-md5 \--enable-eap-mschapv2 --enable-eap-tls --enable-eap-ttls --enable-eap-peap \--enable-eap-tnc --enable-eap-dynamic --enable-eapradius --enable-xauth-eap \--enable-xauth-pam --enable-dhcp --enable-openssl --enable-unity \--enable-certexpire --enable-radattr --enable-tools --enable-attr-sql --enable-mysql --disable-gmp
make
make install

I already have a good config but I need to update it to assign static virtual IP to my clients based on their identity, and all the static virtual IPs are nonsequent IPv6 addresses.
I have configure attr-sql in the list of plugin, but it doesn't work.

[root@localhost strongswan-5.8.4]# /usr/local/sbin/ipsec --version
Linux strongSwan U5.8.4/K3.10.0-1127.el7.x86_64
[root@localhost strongswan-5.8.4]# /usr/local/sbin/ipsec pool
opening database failed.

[root@localhost ~]# ll /usr/local/sbin/
total 1772
-rwxr-xr-x 1 root root   38677 Oct 19 18:46 checkrad
-rwxr-xr-x 1 root root    7734 Nov 13 19:27 ipsec
-rwxr-xr-x 1 root root    3238 Oct 19 18:46 raddebug
-rwxr-xr-x 1 root root 1104472 Oct 19 18:46 radiusd
-rwxr-xr-x 1 root root  186856 Oct 19 18:46 radmin
-rwxr-xr-x 1 root root    7217 Oct 19 18:46 radwatch
-rwxr-xr-x 1 root root    2586 Oct 19 18:46 rc.radiusd
-rwxr-xr-x 1 root root  451232 Nov 13 19:27 swanctl
[root@localhost ~]# ll /usr/local/libexec/ipsec/
total 880
-rwxr-xr-x 1 root root 114624 Nov 13 19:27 charon
-rwxr-xr-x 1 root root  17928 Nov 13 19:27 _copyright
-rwxr-xr-x 1 root root 144256 Nov 13 19:27 pool
-rwxr-xr-x 1 root root 154680 Nov 13 19:27 scepclient
-rwxr-xr-x 1 root root 320768 Nov 13 19:27 starter
-rwxr-xr-x 1 root root  82288 Nov 13 19:27 stroke
-rwxr-xr-x 1 root root  20335 Nov 13 19:27 _updown
-rwxr-xr-x 1 root root  29432 Nov 13 19:27 xfrmi

after execute systemctl restart strongswan-starter, strongswan log like this

Nov 15 16:49:10 localhost systemd: Stopped strongSwan IPsec IKEv1/IKEv2 daemon using ipsec.conf.
Nov 15 16:49:10 localhost ipsec: charon stopped after 400 ms
Nov 15 16:49:10 localhost ipsec: ipsec starter stopped
Nov 15 16:49:10 localhost ipsec: Starting strongSwan 5.8.4 IPsec [starter]...
Nov 15 16:49:10 localhost ipsec: !! Your strongswan.conf contains manual plugin load options for charon.
Nov 15 16:49:10 localhost ipsec: !! This is recommended for experts only, see
Nov 15 16:49:10 localhost ipsec: !! http://wiki.strongswan.org/projects/strongswan/wiki/PluginLoad
Nov 15 16:49:10 localhost systemd: Started strongSwan IPsec IKEv1/IKEv2 daemon using ipsec.conf.
Nov 15 16:49:10 localhost charon: 00[DMN] Starting IKE charon daemon (strongSwan 5.8.4, Linux 3.10.0-1127.el7.x86_64, x86_64)
Nov 15 16:49:11 localhost charon: 00[LIB] openssl FIPS mode(0) - disabled
Nov 15 16:49:12 localhost charon: 00[CFG] loading ca certificates from '/usr/local/etc/ipsec.d/cacerts'
Nov 15 16:49:12 localhost charon: 00[CFG] loading aa certificates from '/usr/local/etc/ipsec.d/aacerts'
Nov 15 16:49:12 localhost charon: 00[CFG] loading ocsp signer certificates from '/usr/local/etc/ipsec.d/ocspcerts'
Nov 15 16:49:12 localhost ipsec: charon (14614) started after 600 ms
Nov 15 16:49:12 localhost charon: 00[CFG] loading attribute certificates from '/usr/local/etc/ipsec.d/acerts'
Nov 15 16:49:12 localhost charon: 00[CFG] loading crls from '/usr/local/etc/ipsec.d/crls'
Nov 15 16:49:12 localhost charon: 00[CFG] loading secrets from '/usr/local/etc/ipsec.secrets'
Nov 15 16:49:12 localhost charon: 00[CFG] attr-sql plugin failed to connect to database
Nov 15 16:49:12 localhost charon: 00[LIB] loaded plugins: charon attr certexpire cmac constraints counters des dhcp dnskey drbg eap-dynamic eap-identity eap-md5 eap-mschapv2 eap-peap eap-tls eap-tnc eap-ttls fips-prf md5 mysql openssl pgp pkcs12 pkcs7 pkcs8 pubkey radattr random rc2 resolve sshkey tnc-tnccs unity vici xauth-eap xauth-generic xauth-pam xcbc nonce aes sha1 sha2 pem pkcs1 curve25519 x509 revocation hmac stroke kernel-netlink socket-default updown
Nov 15 16:49:12 localhost charon: 00[JOB] spawning 16 worker threads
Nov 15 16:49:12 localhost charon: 04[CFG] received stroke: add connection 'gw_eap'
Nov 15 16:49:12 localhost charon: 04[CFG] virtual IP pool too large, limiting to 2409:8805:acce:1::/97
Nov 15 16:49:12 localhost charon: 04[CFG] adding virtual IP address pool 2409:8805:acce:1::/48
Nov 15 16:49:12 localhost charon: 04[CFG] added configuration 'gw_eap'

[root@localhost freeradius-server-release_2_2_7.rsync_from_mysql]# systemctl status strongswan-starter
● strongswan-starter.service - strongSwan IPsec IKEv1/IKEv2 daemon using ipsec.conf
   Loaded: loaded (/usr/lib/systemd/system/strongswan-starter.service; disabled; vendor preset: disabled)
   Active: active (running) since Fri 2024-11-15 16:49:10 CST; 3min 9s ago
 Main PID: 14609 (starter)
   CGroup: /system.slice/strongswan-starter.service
           ├─14609 /usr/local/libexec/ipsec/starter --daemon charon --nofork
           └─14614 /usr/local/libexec/ipsec/charon

Nov 15 16:49:12 localhost.localdomain charon[14614]: 00[CFG] loading crls from '/usr/local/etc/ipsec.d/crls'
Nov 15 16:49:12 localhost.localdomain charon[14614]: 00[CFG] loading secrets from '/usr/local/etc/ipsec.secrets'
Nov 15 16:49:12 localhost.localdomain charon[14614]: 00[CFG] attr-sql plugin failed to connect to database
Nov 15 16:49:12 localhost.localdomain charon[14614]: 00[LIB] loaded plugins: charon attr certexpire cmac constraints co...
Nov 15 16:49:12 localhost.localdomain ipsec_starter[14609]: charon (14614) started after 600 ms
Nov 15 16:49:12 localhost.localdomain charon[14614]: 00[JOB] spawning 16 worker threads
Nov 15 16:49:12 localhost.localdomain charon[14614]: 04[CFG] received stroke: add connection 'gw_eap'
Nov 15 16:49:12 localhost.localdomain charon[14614]: 04[CFG] virtual IP pool too large, limiting to 2409:8805:acce:1::/97
Nov 15 16:49:12 localhost.localdomain charon[14614]: 04[CFG] adding virtual IP address pool 2409:8805:acce:1::/48
Nov 15 16:49:12 localhost.localdomain charon[14614]: 04[CFG] added configuration 'gw_eap'
Hint: Some lines were ellipsized, use -l to show in full.

cat /usr/local/etc/strongswan.conf
charon {
	load_modular = yes
        load = random nonce aes sha1 sha2 pem pkcs1 curve25519 gmp x509 curl revocation hmac stroke kernel-netlink socket-default sqlite attr-sql updown
	plugins {
		include strongswan.d/charon/*.conf
                attr-sql {
                    database = sqlite:///etc/db.d/ipsec.db
                }
	}
}
pool {
    load = sqlite
}
include strongswan.d/*.conf

cat /usr/local/etc/ipsec.conf
config setup
	# strictcrlpolicy=yes
	# uniqueids = no

conn %default
    ikelifetime=29d
    keylife=29d
    rekeymargin=3m
    keyingtries=1
    keyexchange=ikev2
    dpdaction=clear
    mobike=yes
    dpddelay=30s

conn gw_eap
    left=%any
    right=%any
    leftid=%any
    rightid=%any
    leftsubnet=2409:8015:8029:13::/64
    rightsubnet=2409:8805:acce:1::/48
    rightsourceip=2409:8805:acce:1::/48
    leftauth=pubkey
    rightauth=pubkey
    auto=add
    ike=aes128-sha256-modp2048
    esp=aes128-sha1-modp2048
    fragmentation=yes

[root@localhost network]# cat /usr/local/etc/swanctl/swanctl.conf |grep -v "#"
include conf.d/*.conf

I have initialized sqlite db using this script : https://github.com/strongswan/strongswan/blob/master/src/pool/sqlite.sql
[root@localhost local]# sqlite3 /etc/db.d/ipsec.db
SQLite version 3.7.17 2013-05-20 00:56:22
Enter ".help" for instructions
Enter SQL statements terminated with a ";"
sqlite> .schema
CREATE TABLE identities (
  id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
  type INTEGER NOT NULL,
  data BLOB NOT NULL,
  UNIQUE (type, data)
);
...
CREATE TABLE logs (
  id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
  local_spi BLOB NOT NULL,
  signal INTEGER NOT NULL,
  level INTEGER NOT NULL,
  msg TEXT NOT NULL,
  time INTEGER NOT NULL DEFAULT CURRENT_TIMESTAMP
);
sqlite> 

Maybe there is something I didn't understand from documentation...

Thank you for your help !

**Comment by tobiasbrunner** [comment, marked as answer]
> I am currently using strongswan 5.8.4, centos 7.8 x86_64
> I build from code like this:

Why would you use such an ages old version if you build from sources?

Anyway, you are missing the `sqlite` plugin.

**Reply by shenminjie666** [reply]
Thanks for reply, I modify the parameters (add --enable-sqlite) when I build from source, and ipsec pool tool is OK.
./configure --enable-eap-identity --enable-eap-md5 --enable-eap-mschapv2 --enable-eap-tls --enable-eap-ttls --enable-eap-peap --enable-eap-tnc --enable-eap-dynamic --enable-xauth-eap --enable-xauth-pam --enable-dhcp --enable-openssl --enable-unity --enable-certexpire --enable-radattr --enable-attr-sql --enable-sqlite --disable-gmp

By the way, I use 5.8.4 because I want to use ipsec.conf like this, leftauth and rightauth is different.
leftauth=pubkey
rightauth=eap-radius
I worry about that if I use the newest version, the configuration can not support, I can not find the similar config in swanctl.conf

Thank you very much!

**Reply by tobiasbrunner** [reply]
Of course this is supported. Just configure the `auth` setting in the `local` and `remote` sections accordingly. There are many [RADIUS example configs](https://docs.strongswan.org/docs/5.9/config/IKEv2.html) you could have a look at.

**Reply by shenminjie666** [reply]
OK, I will try the new version, thank you very much.
